src/tools/w3dviewer/colorbar.cpp

```diff
@@ -0,0 +1,1170 @@
+    BitBlt(m_deviceContext, 0, 0, m_keyWidth, m_keyHeight, dc, 0, 0, SRCCOPY);
+    SelectObject(m_deviceContext, obj);
+    SelectObject(dc, obj2);
+    DeleteDC(dc);
+    DeleteObject(hbm);
+}
+
+void ColorBarClass::DrawKey(int x, int y)
+{
+    int i1 = 3 * m_keyWidth;
+    int i2 = 3 * m_keyWidth % 4;
+    int i3;
+
+    if (i2 != 0) {
+        i3 = 4 - i2;
+    } else {
+        i3 = 0;
+    }
+
+    if (m_DIBSectionBits != nullptr) {
+        if (m_keyDIBSectionBits != nullptr) {
+            int dest = 3 * x + y * m_stride;
+            int src = 0;
+
+            for (int i = 0; i < m_keyHeight; i++) {
+                for (int j = 0; j < m_keyWidth; j++) {
+                    BYTE red = m_keyDIBSectionBits[src];
+                    BYTE green = m_keyDIBSectionBits[src + 1];
+                    BYTE blue = m_keyDIBSectionBits[src + 2];
+                    src += 3;
+
+                    if (red == BYTE_MAX && green == 0 && blue == BYTE_MAX) {
+                        dest += 3;
+                    } else {
+                        m_DIBSectionBits[dest] = red;
+                        m_DIBSectionBits[dest + 1] = green;
+                        m_DIBSectionBits[dest + 2] = blue;
+                        dest += 3;
+                    }
+                }
+
+                src += i3;
+                dest += m_stride - i1;
+            }
+        }
+    }
+}
+
+int ColorBarClass::FindKey(int x, int y)
+{
+    int len;
+
+    if ((GetWindowLong(m_hWnd, GWL_STYLE) & CLBS_HORIZONTAL) != 0) {
+        len = m_keyWidth;
+    } else {
+        len = m_keyHeight;
+        x = y;
+    }
+
+    int i1 = (len + 1) >> 1;
+    int key = -1;
+    int pos = 1024;
+
+    for (int i = 0; i < m_keyCount; i++) {
+        int i2 = abs(m_keys[i].m_keyPosition - x);
+
+        if (i2 < pos) {
+            pos = i2;
+            key = i;
+        }
+    }
+
+    if (pos >= i1) {
+        return -1;
+    } else {
+        return key;
+    }
+}
+
+LRESULT ColorBarClass::SendNotification(int code, int key)
+{
+    int id = GetWindowLong(m_hWnd, GWL_ID);
+    ColorBarNotify cbn;
+    memset(&cbn, 0, sizeof(cbn));
+    cbn.base.code = code;
+    cbn.keyposition = key;
+    cbn.base.hwndFrom = m_hWnd;
+    cbn.base.idFrom = id;
+    cbn.red = m_keys[key].m_red;
+    cbn.green = m_keys[key].m_green;
+    cbn.blue = m_keys[key].m_blue;
+    cbn.keytime = ((m_sliderMax - m_sliderMin) * m_keys[key].m_keyTime) + m_sliderMin;
+    return GetParent()->SendMessage(WM_NOTIFY, id, reinterpret_cast<LPARAM>(&cbn));
+}
+
+void ColorBarClass::GetSliderRect(RECT *r)
+{
+    *r = m_size;
+    float f1 = (m_sliderPos - m_sliderMin) / (m_sliderMax - m_sliderMin);
+    LONG style = GetWindowLong(m_hWnd, GWL_STYLE);
+
+    if ((style & CLBS_HORIZONTAL) != 0) {
+        int i1 = ((float)(m_size.right - m_size.left) * f1) + m_size.left - 3;
+        r->left = i1;
+        r->right = i1 + 6;
+
+        if ((style & (CLBS_OUTLINEFRAME | CLBS_RAISEDFRAME | CLBS_SUNKENFRAME)) != 0) {
```

Comment:
is this right? as its not merged oddly... & 4 | 3 )

Reply:
GetSliderRect needs same fix in two cases

Reply:
fixed.